### PR TITLE
Fuel quote module

### PR DESCRIPTION
**Pull request by tnttruong**
# CHANGES/NEW FEATURES
- Added a "[+] New Quote" button linked to the `fuel-quote-form` on the same page as `fuel-quote-history`
- Quote 1 and 2 are linked to dummy info and show the current id each page
- Inputs for new quotes are shown on the `fuel-quote-form`

# KNOWN ISSUES
- Quote 1 and Quote 2 buttons don't register new inputted information
- Not linked to database

### CODE REVIEW

File: src/backend/src/database.rs [hunk 1]
```diff
@@ -128,6 +128,12 @@ pub async fn authenticate_user(pool: &Pool<Sqlite>, username: &str, password: &s
 
 }
 
+/*
+pub async fn store_quote(pool: &Pool<Sqlite>, gallons: &i32, address: &str, date: &str) -> Result<String, Box<dyn Error>>{
```
Comment (Brod8362) [inline src/backend/src/database.rs:132]: I know this is commented out, but let's remove it for now - it is not relevant to the scope of assn 3.

File: src/backend/src/main.rs [hunk 2]
```diff
@@ -31,6 +31,9 @@ async fn rocket() -> _ {
                 routes::logout,
                 routes::profile,
                 routes::quote,
+                routes::quote_id,
+                routes::quote_request,
```
Comment (Brod8362) [inline src/backend/src/main.rs:35]: :+1:

File: src/backend/src/routes.rs [hunk 3]
```diff
@@ -123,7 +122,53 @@ pub fn quote() -> Template {
     )
 }
 
-#[get("/page/quote_history")]
+#[derive(FromForm)]
+pub struct QuoteData {
```
Comment (Brod8362) [inline src/backend/src/routes.rs:126]: Make sure that fields in this struct match the names of the form being submitted.
That's how Rockets how to properly de-serialize the field.

File: src/backend/src/routes.rs [hunk 4]
```diff
@@ -123,7 +122,53 @@ pub fn quote() -> Template {
...
+    }
+    */
+
+    //TODO: have this happen once quote_request is OK
+    return String::from("<p> Fuel Quote Submitted </p>")
+}
+
+#[get("/page/quote/<id>")]
+pub fn quote_id(id: &str) -> Template {
+    Template:: render(
```
Comment (Brod8362) [inline src/backend/src/routes.rs:150]: This is good. If we can, it would be nice to have the rendered template at least display the ID (this will help in unit testing)

File: src/backend/src/routes.rs [hunk 5]
```diff
@@ -123,7 +122,53 @@ pub fn quote() -> Template {
...
+        }
+    )
+}
+
+#[post("/page/quote/<id>", data="<form>")]
+pub async fn quote_submit(id: &str, form: Form<QuoteData>) -> Template {
+    let form_input = form.into_inner();
+    Template:: render(
+        "fuel_quote",
+        context!{
```
Comment (Brod8362) [inline src/backend/src/routes.rs:163]: This function is redundant and should not exist, the functionality is handled by `POST /page/quote`

File: src/backend/templates/fuel_quote_form.html.tera [hunk 6]
```diff
@@ -61,12 +61,20 @@
             <p>Total Amount Due: 0.00 USD</p>
             <!-- total is a noneditable numeric value calculated (gallons * price per gallon)-->
         </section>
-
         <br>
-        <button type="submit" hx-get="/fuel_quote" hx-trigger="click" hx-target="#fuel-quote-form">
+
+        <!-- FOR TEMPLATE APPLICATION -->
```
Comment (Brod8362) [inline src/backend/templates/fuel_quote_form.html.tera:66]: I like the comments where the template fields are.

I think this is something we should adopt throughout the whole project - if there's a template field, a brief comment explaining what it's for/what is meant to be there is helpful.

Good job :+1:

File: src/backend/templates/fuel_quote_history.html.tera [hunk 7]
```diff
@@ -1,70 +1,12 @@
...
-            <td>5113 Rainflower Circle S, League City, TX 77573</td>
-            <td>2023-11-15</td>
-            <td>2.00</td>
-            <td>200.00</td>
-        </tr>
-    </table>
-</div>
+    <div class="tab">
+        <a hx-get="/page/quote" hx-target="#quote-content">[+] New Quote</a>
+        <a hx-get="/page/quote/1" hx-target="#quote-content">Quote 1</a>
```
Comment (Brod8362) [inline src/backend/templates/fuel_quote_history.html.tera:6]: Looks good here - eventually, there will be a loop here that generates the content automatically (but that must wait for database implementation)

File: src/backend/templates/home.html.tera [hunk 8]
```diff
@@ -12,23 +12,21 @@
...
-        <li><a hx-get="/page/profile" hx-swap="innerHTML" hx-target="#content">Profile</a></li>
-        <li><a hx-get="/logout" hx-swap="innerHTML" hx-target="#content"></a></li>
+        <li><a hx-get="/page/quote/history" hx-swap="innerHTML" hx-target="#title-form">Fuel Quote History</a></li>
+        <li><a hx-get="/page/quote" hx-swap="innerHTML" hx-target="#title-form">Generate Fuel Quote</a></li>
+        <li><a hx-get="/page/profile" hx-swap="innerHTML" hx-target="#title-form">Profile</a></li>
+        <li><a hx-get="/logout" hx-swap="innerHTML" hx-target="#title-form"></a></li>
     </ul>
     {% else %}
-    <a id="account" hx-get="/login" hx-swap="innerHTML" hx-target="#content">Login / Register</a> 
+    <a id="account" hx-get="/login" hx-swap="innerHTML" hx-target="#content-form">Login / Register</a>
```
Comment (Brod8362) [inline src/backend/templates/home.html.tera:21]: As mentioned on #10, I don't agree with the main content div being called `#content-form`. It is *not* a form, it is a div/section etc. 

The ID of an element should be as accurate of a representation as possible of what the element is actually for, and `#content-form` is misleading.

File: src/backend/templates/home.html.tera [hunk 9]
```diff
@@ -12,23 +12,21 @@
...
     <!--<img id="logo" src="placeholder.png" alt="logo">  FIND OUT HOW TO MOVE THIS -->
 </header>
 
 <body>
-    <div id="bottom_half">	
-        <div id="content">
-            
-        </div>
+    <div id="content-form">	
+        <div id="title-form"> </div>
```
Comment (Brod8362) [inline src/backend/templates/home.html.tera:29]: Same here - the title isn't a form, so it should not be referred to as one.

File: src/backend/templates/home.html.tera (hunk 8 above)
Comment (Brod8362) [reply src/backend/templates/home.html.tera:21]: A name like `#main-content` might be more fitting if we don't want to use `#content`.

File: src/backend/templates/login_page.html.tera [hunk 10]
```diff
@@ -1,58 +1,58 @@
...
+    <style>
+        .form-sect{
+            display: flex;
+            align-items: center;
+        }
+    </style>
+</head>
+<body>
+    <div>
+        <h1>Welcome!</h1>
```
Comment (Brod8362) [inline src/backend/templates/login_page.html.tera:13]: Not much to say here - just formatting changes (and the `#content-form` thing again)

**Review by Brod8362 — CHANGES_REQUESTED** [review]
Please review the comments & make changes accordingly.

This PR does not have any unit tests, so we need to make sure those are implemented.

Every function written must have at *least* two unit tests (it is OK and even recommended that there are more) - one that ensures it breaks when it should, and one that ensures it works when it should. Unit tests must also try to ensure every code path is taken at least once.

This helps in finding bugs and unexpected behavior. Try to think of edge cases where things might break, and make sure they either

A - break in a safe way 
or
B - are fixed, in the event that they are bugs.

File: src/backend/src/routes.rs [hunk 11]
```diff
@@ -141,29 +142,28 @@ pub async fn quote_request(pool: &State<Pool<Sqlite>>, form: Form<QuoteData>) ->
...
-    //TODO: have this happen once quote_request is OK
-    return String::from("<p> Fuel Quote Submitted </p>")
-}
+    //TODO: return submission confirmation upon OK
+    //return String::from("<p> Fuel Quote Submitted </p>")
 
-#[get("/page/quote/<id>")]
-pub fn quote_id(id: &str) -> Template {
+    //renders form results FOR NOW until database implement
+    let form_inp = form.into_inner();
```
Comment (Brod8362) [inline src/backend/src/routes.rs:149]: `.into_inner()` should not be necessary here, off the top of my head. You can access the fields without it (you will probably need to borrow (`&`))

File: src/backend/src/routes.rs [hunk 12]
```diff
@@ -226,4 +229,11 @@ mod tests {
         //TODO: need to figure out how to instantiate a CookieJar from scratch and call the correct request method
         todo!()
     }
+
+    #[tokio::test]
+    async fn test_index() {
```
Comment (Brod8362) [inline src/backend/src/routes.rs:234]: We don't need a unit test for the index, test the code that you've written (e.g the profile module)

(I understand this is probably just temporary)

**Review by Brod8362 — CHANGES_REQUESTED** [review]
Looks generally good, see small change requests.

Once the unit tests are in and look good, this is ready to merge
